ratis-logservice/src/main/java/org/apache/ratis/logservice/api/LogReader.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.logservice.api;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * Synchronous client interface to read from a LogStream.
+ */
+public interface LogReader extends AutoCloseable {
+
+  /**
+   * Seeks to the position before the record at the provided {@code offset} in the LogStream.
+   *
+   * @param offset A non-negative, offset in the LogStream
+   * @return A future for when the operation is completed.
+   */
+  void seek(long recordId) throws IOException;
```

Comment:
Seeking by record id is expensive to implement. One has to support separate index {recordid -> file offset}. This should be not ID, but generic "offset", which maps 1 -> 1 to file offset. The only use case, where we need positional reads (at least in logs) is to resume sequential reading from the last read offset (HBase replication as an example).
We have to separate logs from message queues - they are totally different by their purpose.

Reply:
That's why most of the implementations of distributed logs (pub/sub model) actually are using segments/slices/whatever they call it and why usually they are relatively small. Usually, there is a separate index that maintains the map between id and segment OR it keeps start id for the segment in its name.  What do you mean by 'separate logs from message queues'?

Reply:
One more thing. Actually, such seek operation should be already implemented under the hood because it's required for quick sync up for lagging quorum nodes due network problems/restarts ( didn't dig into the code yet, but almost 100% sure why segments were implemented there). All we need is to provide a facade that can be used outside.

Reply:
I meant what I meant: logs are not a message queues. Period. There is no need for random access in log file to grab a particular log record, but there is such a need in message queue. As for seeking support in Raft log - there is no such requirement to implement this efficiently, quite contrary, afair, protocol describes very stupid and lazy algo of synching two logs: leader goes backward one record at a time and compares it with a lagging follower records. 
Do not bring please functionality we (Log service users) are not asking for.

Reply:
Moving to a specific point in a LogStream is absolutely a necessary API call. We must be able to resume reading from a known point in the Log. HBase Replication needs that functionality.

A naive implementation can read all records from the LogStream to know where to start reading. A slightly more advanced implementation can use the underlying segment files to find the right segment file to read from. An advance implementation could consult an index which we explicitly maintain over all segment files.

We haven't touched on implementation of how the LogReader class would negotiate this call with the back-end StateMachine (that depends on what the StateMachine and/or Ratis APIs look like, I think).

Reply:
>> Moving to a specific point in a LogStream is absolutely a necessary API call

I think everyone here agree with you. The issue is keeping separate indexes for records. I do not know how this was implemented in Ratis, but in any case, maintaining index eats CPU and IO and slow everything down. If it is there already - fine, we can reuse it. To resume HBase replication  - offset in a log stream (not record id) would suffice (in our use case, of course).

Reply:
Ratis has a directory in which is stores many LogSegments. Each of these are ordered with an Index, per the Raft protocol. Ratis can find the LogSegment containing the record at an Index using binary search and then does a linear search over that LogSegment to find the specific ID. I think that's sufficient for what we need in a first-pass.

I don't think we need to support a "recordId" that is different than the Raft Index. Just trying to give it a better name.